SourceCode/include/tracer/Ray.h

```diff
@@ -2,17 +2,30 @@
 
 #include <optional>
 
-#include "Utils.h"
 #include "Vector.h"
 #include "tracer/Triangle.h"
 
-struct Ray {
-  constexpr static const float FLOAT_EPSILON = 1e-8;
+enum RayType { Primary, Shadow };
+
+template <RayType T = Primary>
```

Comment:
Instead of using template here you could add flags member, because other properties for the ray will be needed